File: 0214 Shortest Palindrome/0214.cpp
class Solution {
public:
    string shortestPalindrome(string s) {
        int i = 0;
        int n = s.size();

        for (int j = n-1; j >= 0; j--){
            if (s[i] == s[j]){
                i++;
            }
        }

        if (i == n){
            return s;
        }

        string mid = s.substr(0, i);
        string r = s.substr(i);
        string l = r;
        reverse(l.begin(), l.end());

        return l + shortestPalindrome(mid) + r;
    }
};

File: 0214 Shortest Palindrome/0214.py
class Solution:
    def shortestPalindrome(self, s: str) -> str:
        i = 0
        n = len(s)

        for j in range(n):
            if s[i] == s[n-1-j]:
                i += 1

        if i == n:
            return s

        return s[i:][::-1] + self.shortestPalindrome(s[:i]) + s[i:]
